frappe_graphql/utils/subscriptions.py

```diff
@@ -0,0 +1,304 @@
+from frappe.utils import now_datetime, get_datetime
+
+from frappe_graphql import get_schema
+from frappe_graphql.utils.resolver import default_field_resolver
+
+"""
+Implemented similar to
+https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md
+
+Server --> Client Message Types:
+- GQL_DATA
+- GQL_COMPLETE
+
+Only the above two are implemented as of now. Once we have a mechanism for
+SocketIO -> Python communication in frappe, we can implement the complete spec
+which includes types like:
+- GQL_START
+- GQL_STOP
+- GQL_CONNECTION_ACK
+- GQL_CONNECTION_KEEP_ALIVE
+"""
+
+
+def setup_subscription(subscription, info: GraphQLResolveInfo, variables, complete_on_error=False):
+    """
+    Set up a frappe task room for the subscription
+    Args:
+        subscription: The name of the subscription, usually the field name itself
+        info: The graphql resolve info
+        variables: incoming variable dict object
+        complete_on_error: Stop / Send Completed Event on GQL Error
+
+    Returns:
+        Subscription info, including the subscription_id
+    """
+    excluded_field_nodes = filter_selection_set(info)
+    variables = frappe._dict(variables)
+    subscription_id = frappe.generate_hash(f"{subscription}-{frappe.session.user}", length=8)
+
+    subscription_data = frappe._dict(
+        subscribed_at=now_datetime(),
+        last_ping=now_datetime(),
+        variables=variables,
+        subscription_id=subscription_id,
+        selection_set=excluded_field_nodes,
+        user=frappe.session.user,
+        complete_on_error=complete_on_error
+    )
+
+    frappe.cache().hset(
+        get_subscription_redis_key(subscription), subscription_id, subscription_data)
+
+    return frappe._dict(
+        subscription_id=subscription_id
+    )
+
+
+def get_consumers(subscription):
+    """
+    Gets a list of consumers subscribed to a particular subscription
+    Args:
+        subscription: The name of the subscription
+
+    Returns:
+        A list of consumers with their subscription info
+    """
+    redis_key = get_subscription_redis_key(subscription)
+    consumers = frappe.cache().hgetall(redis_key)
+    return consumers.values()
+
+
+def notify_consumer(subscription, subscription_id, data):
+    consumer = frappe.cache().hget(
+        get_subscription_redis_key(subscription),
+        subscription_id
+    )
+    if not consumer:
+        return
+
+    original_user = frappe.session.user
+    frappe.set_user(consumer.user)
+
+    execution_data = gql_transform(subscription, consumer.selection_set, data or frappe._dict())
+    response = frappe._dict(
+        type="GQL_DATA",
+        id=subscription_id,
+        payload=execution_data
+    )
+
+    room = get_task_room(subscription_id)
+    emit_via_redis(
+        event=subscription,
+        message=response,
+        room=room
+    )
+
+    if len(execution_data.get("errors") or []):
+        log_error(consumer, execution_data)
```

Comment:
@fahimalizain This function is missing the 1 positional argument

Reply:
@assemmarwan  resolved 👍🏼